### ISSUE TITLE
[Discussion] Azure Functions Runtime 2.0.11888 release

### ISSUE THREAD

**Issue by alrod** [Azure/app-service-announcements-discussions#54]
Discussion issue for the Azure Functions runtime release 1.0.11866

**Comment by JohnReeson** [comment]
So the version is 2.0.11888, right?

**Comment by alrod** [comment]
@JohnReeson right

**Comment by davidebbo** [comment]
I just fixed the title.

**Comment by TurtleBeach** [comment]
I have deployed a function in west central US that is running the 2.0.11888 runtime. It was deployed with Anonymous access (as I often do for a quick check), but when I click the Manage link to change access to Function, a pop-up Error box appears: Error: Unauthorized - /api/passthrough. No function keys appear, and if I try to add one the same error appears.
If I click the Monitor link, a different error appears: Error: The call to get Host information failed. In this case, the background page first shows the "classic" logging view (which is what I want), but after the error appears, the "Configure Application Insights" page appears, with no option to use Classic logging.
In short, the deployment is not yet ready for prime time, though the function does work when Anonymous access is permitted.

**Comment by alrod** [comment]
@TurtleBeach, It's known portal issue when you set WEBSITE_USE_PLACEHOLDER=1. Function runtime itself should work fine in this case. You can try:
Restart your function app and then do your changes on the portal. Or set back to WEBSITE_USE_PLACEHOLDER=0, do your changes on the portal, set  WEBSITE_USE_PLACEHOLDER=1.

**Comment by TurtleBeach** [comment]
@alrod restarting the function with WEBSITE_USE... = 1 did not help (I had previously stopped and started several times, but I tried again). HOWEVER, when I set WEBSITE_USE... to 0 and restarted, I was able to change function access and provide the key needed by callers of the function. I then reset WEBSITE_USE.. to 1, restarted again, and the function works. In a quick test, I there seems to be a small decrease in start-up from cold, but a noticeably larger decrease the second time the function is called, at least within a couple of minutes. A big improvement.
Jack

**Comment by TurtleBeach** [comment]
But please get the Azure deployment issues sorted. There are enough things to get right with my own function deployment - dealing with Azure issues is painful on top of that.

**Comment by alrod** [comment]
This runtime deployment has been completed.

**Comment by jamesaarondevlin** [comment]
hey @alrod - does this mean we can flip our function apps over to 64-bit now or will there be a separate announcement regarding that?

**Comment by alrod** [comment]
@jamesaarondevlin, there will be a separate announcement. Probably after next Azure Functions Runtime 2.0 deployment you will be able to use 64-but again.

**Comment by willmorgan** [comment]
On the 64-bit change front, we noticed that running `beta` and restarting would update the Function App's version. As we were running in 64-bit mode, then that update actually downdated us to `1.x` and thus break our code quite badly 😅

This update behaviour obviously makes me feel quite uneasy, so will `2.0.11887` be available for a while until we can move to the next release, presumably `2.0.11889`?

**Comment by davidebbo** [comment]
@willmorgan the behavior I would expect when setting it to 'beta' is:

- Previous build: you could set the app to 64 bit, but it would still end up running in 32 bit 
- Now that we released `2.0.11888`: setting it to 64 bit breaks things entirely

There is no case where setting it to beta should cause it to use 1.x.

**Comment by TurtleBeach** [comment]
2.0.11888 appears to be running on different MS servers from previous releases.
With the earlier runtimes, response headers indicated that the server was Kestrel. Now response headers indicated the server is IIS / 10.0. More disturbingly, the response behavior has changed.
With the earlier runtime, a response JSON payload was returned (if of course intended) regardless of the HTTP response code (in other words, a success or accepted code could have a payload, and an error return could have a payload).
Now, the payload is still returned with success codes, but the payload for error returns (e.g., 400, 500, 460 as a custom code) is NOT returned. In the case of HTTP error codes, only the error code, and a canned string roughly describing the error is returned (e.g. 404 returns something like the resource was not found).
It seems this anomaly is being caused by a "Custom Error Module", since for the HTTP code of 460, the return phrase is "The custom error module does not recognize this error."
THIS IS A SERIOUS PROBLEM for me - I have entire system of functions which relies on information being available in return JSON payloads when errors occur. And in general, there is no reason for some kind of "custom error module" to be intercepting and filtering returns from functions.
I have raised this issue on MSDN, through my portal support, and now here.
I would appreciate a response.

**Comment by davidebbo** [comment]
@TurtleBeach Please see the announcement (https://github.com/Azure/app-service-announcements/issues/117), which captures this issue.

**Comment by TurtleBeach** [comment]
David, thanks. It would have been helpful to have that sentence 3 days ago. There are two or three threads you guys are using (and the last time I looked the reference thread was closed).
Also, can I assume that since it is a known issue it will be fixed?
Jack (TurtleBeach)

**Comment by davidebbo** [comment]
@TurtleBeach We only discovered this issue 2 days ago, so it was not something we knew about pre-deployment. Yes, it will definitely be fixed. This scenario clearly needs to work. Sorry for the inconvenience.

**Comment by willmorgan** [comment]
@davidebbo We were running beta in one of our application environments and the runtime crashed, and was unable to restart.

I've got us running on `2.0.11857.0-alpha` for now. When does this version get sunsetted so we know when we need to update to the next (hopefully stable) version?

**Comment by davidebbo** [comment]
@willmorgan did you see the [announcement](https://github.com/Azure/app-service-announcements/issues/117)? Make sure you're not running into the 64 bit issue.

The older version will get removed when we feel all relevant issues with the newer one have been addressed.

**Comment by willmorgan** [comment]
@davidebbo Yes, I did. We hadn't touched that Function App for a few days. If the runtime version auto upgrades to 2.0.11888, then should it not gracefully set 32 bit?

**Comment by davidebbo** [comment]
@willmorgan I think you are misreading the announcement. If you're set to 64 bit, you need to change it to 32 bit.

**Comment by willmorgan** [comment]
Hi @davidebbo - OK, sure, that's a given. My point is that the configuration we had was:

`FUNCTIONS_EXTENSION_VERSION=beta`
`use32BitWorkerProcess=false` (just grabbed that out of our ARM template)

If this version was released with a known issue which breaks 64 bit apps, then putting a hold on upgrading would have been a nice to have. Alternatively, simply downgrading the site to use 32 bit instead of 64 bit architecture would be workable.

In this case I've followed your advice, so thanks for that, but my team would like to see a smoother upgrade process.

We can't run on `beta` forever, we don't have a date from Microsoft about when `2.0.11857.0-alpha` will be removed, and we don't have a date on when we can expect the next release that fixes this. So some solid advice there would be great 👍

**Comment by davidebbo** [comment]
@willmorgan let me clarify: even though you had your app set to 64 bits before, **it was actually running as 32 bits** (there has never been 64 bit support). The only change is that now this setting breaks this app, hence the announcement ahead of time. The right action here is not to switch to the older version, but simply to switch to 32 bit.

**Comment by ygnr** [comment]
We are also facing the same issue(Custom Error Module) as mentioned by @TurtleBeach. Any timeline when this will be fixed?

**Comment by TurtleBeach** [comment]
I have a mix of Java functions, some running 2.0.11857 because I need the JSON response payload, and a couple running 2.0.11888. Because I was seeing exceptions due to a function never fully starting before the runtime threw a System timeout exception, I changed the function time to 10 minutes. That seems to have helped with the Functions running on 2.0.11857, but I still periodically get a System timeout exception after a "cold Start", with 2.0.11888 and WEBSITE_USE_PLACEHOLDER set to 1. Is this behavior (functions timing out before they run because the runtime is too slow starting) something that MS is working on? This is aggravating, and I don't think there is anything I can do (I am open to suggestions - or an answer to my MSDN query about making Java functions run on a VM that is already running on an AppService plan). Also, I can provide information that exists in AppInsights if that would be helpful.
Thanks

**Comment by davidebbo** [comment]
We are hoping to release a fix to the Custom Error Module issue next week. @TurtleBeach for your other issue, can you open an issue in https://github.com/Azure/azure-functions-host/? This will need more investigation, as I'm not sure it has come up.

**Comment by TurtleBeach** [comment]
@alrod I have been running a queue-triggered function for ~ 2 weeks on 2.0.11888, with WEBSITE_USE_PLACEHOLDER = 1. After sorting out AppInsight issues, I have a page that monitors all my critical functions (Http triggered functions are still on 2.0.11857 due to the Custom Error Module issue). It appears that what WEBSITE_USE_PLACEHOLDER = 1 does is fire the function at whatever the timeout limit is, with some then additional back-off time (e.g., if timeout is 10 minutes, after a successful invocation, if no other invocation occurs within 10 minutes, the function attempts to start - even with no message in the associated queue, and after 10 minutes a timeout exception occurs. Then 10+ minutes later (again, no intervening "real" invocation), something triggers the function again (and again a timeout exception is thrown). From the high-level standpoint of my system, the function is responsive to "real" requests that occur far apart; HOWEVER, the issue is that all the TIMEOUT exceptions now clutter up the AppInsights log, making it virtually useless when trying to find a "real" function error. Surely the source of the timeout (more specifically, the source of the function trigger) is available to the interrupt handler, and if the function was triggered by whatever code is executed as the result of WEBSITE_USE_PLACEHOLDER = 1, the timeout should be silently ignored (i.e., NOT logged by AppInsights).
Thanks.
TurtleBeach AKA -jack-
